packages/app/src/specs/banners/TrackedBanner.vue

```diff
@@ -43,15 +53,37 @@ mutation TrackedBanner_SetProjectState($value: String!) {
 }
 `
 
+gql`
+mutation TrackedBanner_recordBannerSeen($campaign: String!, $messageId: String!, $medium: String!, $cohort: String) {
+  recordEvent(campaign: $campaign, messageId: $messageId, medium: $medium, cohort: $cohort)
+}
+`
+
 const props = withDefaults(defineProps<TrackedBannerComponentProps>(), {})
 
 const emit = defineEmits<TrackedBannerComponentEmits>()
 
 const stateQuery = useQuery({ query: TrackedBanner_ProjectStateDocument })
 const setStateMutation = useMutation(TrackedBanner_SetProjectStateDocument)
+const reportSeenMutation = useMutation(TrackedBanner_RecordBannerSeenDocument)
+const bannerInstanceId = ref(nanoid())
+
+const recordBannerShown = useDebounceFn(({ campaign, medium, cohort }: EventData) => {
+  reportSeenMutation.executeMutation({
+    campaign,
+    messageId: bannerInstanceId.value,
+    medium,
+    cohort: cohort || null,
+  })
+}, 1000)
 
 watchEffect(() => {
   if (props.modelValue) {
+    if (!props.hasBannerBeenShown && props.eventData) {
```

Comment:
Non-blocking: It seems like the debounce here could be avoided if we directly watch only the `modelValue`. The purpose of the `watchEffect` is that you _want_ to run an effect for every reactive value referenced inside that changes, so the `watchEffect` into a debounced function is a little unintuitive to me.

Here's how it would look without:

```js
const recordBannerShown = ({ campaign, medium, cohort }: EventData) => {
  reportSeenMutation.executeMutation({
    campaign,
    messageId: bannerInstanceId.value,
    medium,
    cohort: cohort || null,
  })
}

watch(() => props.modelValue, (newVal) => {
  if (newVal) {
    if (!props.hasBannerBeenShown && props.eventData) {
      // We only want to record the banner being shown once per user, so only record if this is the *first* time the banner has been shown
      recordBannerShown(props.eventData)
    }

    updateBannerState('lastShown')
  }
}, {
  immediate: true,
})
```

I do like the debounce test though, since it is quite important to not repeatedly send these. Maybe I'm missing something else about the watchEffect though.